### PR TITLE
Add nRF Cloud A-GPS support

### PR DESCRIPTION
**Pull request by jtguggedal**
This PR depends on #1956, and is rebased on top of it. Only the two latest commits are relevant for review here.

---

Add support for requesting A-GPS data from nRF Cloud and
processing the response.

Signed-off-by: Jan Tore Guggedal <jantore.guggedal@nordicsemi.no>

---

Adds a minimal sample showing hos nRF Cloud and A-GPS can be
used together.
Fetches A-GPS data upon connection to cloud, and starts GPS
when the data is stored to the modem. Searches for fix for
a maximum of 2 minutes, and restarts search every 4 minutes.
Button press sends a short message to cloud. Long press, 3 secs,
reboots the device.
Sending {"reboot":true} from the cloud to the device also reboots.

Signed-off-by: Jan Tore Guggedal <jantore.guggedal@nordicsemi.no>

### CODE REVIEW

File: samples/nrf9160/nrf_cloud_agps/Kconfig [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+config CLOUD_BACKEND
+	string "String that selects the cloud backend to be used"
+	default "NRF_CLOUD"
+
+config CLOUD_MESSAGE
+	string "Custom message published periodically to cloud"
+	default "{\"state\":{\"reported\":{\"message\":\"Hello Internet of Things!\"}}}"
+
+config POWER_SAVING_MODE_ENABLE
+	bool "Request PSM from cellular network"
```
Comment (SebastianBoe) [inline samples/nrf9160/nrf_cloud_agps/Kconfig]: Most boolean options enable things. It is best practice to rename to POWER_SAVING_MODE (or perhaps add a prefix to disambiguate).

CLOUD_POWER_SAVING_MODE

Comment (jtguggedal) [reply samples/nrf9160/nrf_cloud_agps/Kconfig]: Agreed, I'll rename to `LTE_PSM`, `LTE_POWER_SAVING_MODE` or something along those lines to specify what sort of power saving this enables

**Comment by jtguggedal** [comment]
Updated according to review comment. Documentation will follow in a separate PR.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c [hunk 2]
```diff
@@ -0,0 +1,561 @@
...
+
+	for (size_t i = 0; i < NRF_CLOUD_AGPS_MAX_SV_TOW; i++) {
+		dst->sv_tow[i].flags = src->time_and_tow->sv_tow[i].flags;
+		dst->sv_tow[i].tlm = src->time_and_tow->sv_tow[i].tlm;
+	}
+
+	return 0;
+}
+
+int agps_send_to_modem(struct nrf_cloud_apgs_element *agps_data)
```
Comment (rlubos) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c]: `static`?

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c [hunk 3]
```diff
@@ -0,0 +1,561 @@
...
+					= GPS_AGPS_KLOBUCHAR_CORRECTION,
+	[NRF_GNSS_AGPS_NEQUICK_IONOSPHERIC_CORRECTION]
+					= GPS_AGPS_NEQUICK_CORRECTION,
+	[NRF_GNSS_AGPS_GPS_SYSTEM_CLOCK_AND_TOWS]
+					= GPS_AGPS_GPS_SYSTEM_CLOCK_AND_TOWS,
+	[NRF_GNSS_AGPS_LOCATION]	= GPS_AGPS_LOCATION,
+	[NRF_GNSS_AGPS_INTEGRITY]	= GPS_AGPS_INTEGRITY,
+};
+
+void agps_print_enable(bool enable)
```
Comment (rlubos) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c]: Is this function actually used anywhere?

File: subsys/net/lib/nrf_cloud/include/nrf_cloud_agps.h [hunk 4]
```diff
@@ -0,0 +1,37 @@
+/*
```
Comment (rlubos) [inline subsys/net/lib/nrf_cloud/include/nrf_cloud_agps.h]: I see that this header is used in samples, shouldn't it be moved into `include/net` then? That's a place for public headers.

File: applications/asset_tracker/src/main.c [hunk 5]
```diff
@@ -258,6 +262,34 @@ void cloud_error_handler(int err)
 	error_handler(ERROR_CLOUD, err);
 }
 
+static void send_agps_request(struct k_work *work)
+{
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+	int err;
+	static s64_t last_request_timestamp;
+
+	ARG_UNUSED(work);
```
Comment (rlubos) [inline applications/asset_tracker/src/main.c]: I guess you could move this line out of the `#ifdef` block.

File: samples/nrf9160/nrf_cloud_agps/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+#
+
+cmake_minimum_required(VERSION 3.8.2)
+
+include($ENV{ZEPHYR_BASE}/../nrf/cmake/boilerplate.cmake NO_POLICY_SCOPE)
+
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
+project(cloud_client)
```
Comment (rlubos) [inline samples/nrf9160/nrf_cloud_agps/CMakeLists.txt]: `project(nrf_cloud_agps)`

File: samples/nrf9160/nrf_cloud_agps/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+# NORDIC SDK APP START
+target_sources(app PRIVATE src/main.c)
+# NORDIC SDK APP END
+
+zephyr_include_directories(src)
```
Comment (rlubos) [inline samples/nrf9160/nrf_cloud_agps/CMakeLists.txt]: I don't think it's needed, there're no headers at all in there.

File: samples/nrf9160/nrf_cloud_agps/Kconfig [hunk 8]
```diff
@@ -0,0 +1,28 @@
+#
+# Copyright (c) 2020 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+#
+
+menu "nRF Cloud and A-GPS Sample Settings"
+
+config CLOUD_BACKEND
```
Comment (rlubos) [inline samples/nrf9160/nrf_cloud_agps/Kconfig]: It's nrf_cloud sample, I don't think we have a choice here.

File: samples/nrf9160/nrf_cloud_agps/src/main.c [hunk 9]
```diff
@@ -0,0 +1,462 @@
...
+static struct device *gps_dev;
+static u64_t start_search_timestamp;
+static u64_t fix_timestamp;
+static struct k_delayed_work cloud_update_work;
+static struct k_delayed_work gps_start_work;
+static struct k_delayed_work reboot_work;
+
+static void gps_start_work_fn(struct k_work *work);
+
+static void cloud_update_work_fn(struct k_work *work)
```
Comment (rlubos) [inline samples/nrf9160/nrf_cloud_agps/src/main.c]: I wonder if those periodic updates should be a part of this sample. All in all we're exposing agps functionaly here. Up for discussion.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c (hunk 3 above)
Comment (jtguggedal) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c]: It's used in a test app that I made to be able to see all A-GPS data. It won't go public, so I could probably remove it, along with `nrf_cloud_agps_utils.c`?

File: subsys/net/lib/nrf_cloud/include/nrf_cloud_agps.h (hunk 4 above)
Comment (jtguggedal) [reply subsys/net/lib/nrf_cloud/include/nrf_cloud_agps.h]: Agreed, moved now

File: samples/nrf9160/nrf_cloud_agps/Kconfig (hunk 8 above)
Comment (jtguggedal) [reply samples/nrf9160/nrf_cloud_agps/Kconfig]: Too much copy-pasting... Removed the option.

File: samples/nrf9160/nrf_cloud_agps/src/main.c (hunk 9 above)
Comment (jtguggedal) [reply samples/nrf9160/nrf_cloud_agps/src/main.c]: Yeah, they're not really used anyway. It only sends messageupon button press. Converted to regular function, not work item.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c [hunk 10]
```diff
@@ -0,0 +1,561 @@
...
+ */
+
+#include <zephyr.h>
+#include <device.h>
+#include <drivers/gps.h>
+#include <net/socket.h>
+#include <nrf_socket.h>
+
+#include <modem_info.h>
+#include <nrf_cloud_agps.h>
```
[the remark this answers is not in the document]
Comment (jtguggedal) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c]: They are indeed, thanks

**Comment by jtguggedal** [comment]
@rlubos Addressed your comments.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c (hunk 3 above)
Comment (rlubos) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c]: Well, I'm not sure if we should remove it, it seems useful. Perhaps we should add a Kconfig switch to the library to enable debug mode?

**Pull request merged**